### PR TITLE
Migrate to kotlin

### LINKED ISSUE #3: The GIT repository has JAVA code, the
https://developer.android.com/codelabs/data-providers  is written for Kotlin code and refers to GIT
 https://github.com/googlecodelabs/data-providers.git which is JAVA.

Nobody reviewed this project ?
This is so disappointing...

### PR DESCRIPTION
**Pull request by JolandaVerhoef**

### CODE REVIEW

File: 1-base/src/main/java/com/example/android/wearable/complicationsdataprovider/ComplicationTapBroadcastReceiver.kt [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+        // Retrieve data via SharedPreferences.
+        val preferenceKey = getPreferenceKey(provider, complicationId)
+        val sharedPreferences =
+            context.getSharedPreferences(COMPLICATION_PROVIDER_PREFERENCES_FILE_KEY, 0)
+        var value = sharedPreferences.getInt(preferenceKey, 0)
+
+        // Update data for complication.
+        value = (value + 1) % MAX_NUMBER
+        sharedPreferences.edit { putInt(preferenceKey, value) }
```
Comment (codingjeremy) [inline 1-base/src/main/java/com/example/android/wearable/complicationsdataprovider/ComplicationTapBroadcastReceiver.kt]: Don't you still have to do an `apply()` here?

File: 2-short-data/src/main/java/com/example/android/wearable/complicationsdataprovider/ComplicationTapBroadcastReceiver.kt [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+        // Retrieve data via SharedPreferences.
+        val preferenceKey = getPreferenceKey(provider, complicationId)
+        val sharedPreferences =
+            context.getSharedPreferences(COMPLICATION_PROVIDER_PREFERENCES_FILE_KEY, 0)
+        var value = sharedPreferences.getInt(preferenceKey, 0)
+
+        // Update data for complication.
+        value = (value + 1) % MAX_NUMBER
+        sharedPreferences.edit { putInt(preferenceKey, value) }
```
Comment (codingjeremy) [inline 2-short-data/src/main/java/com/example/android/wearable/complicationsdataprovider/ComplicationTapBroadcastReceiver.kt]: `apply()`?

File: 3-trigger-updates/src/main/java/com/example/android/wearable/complicationsdataprovider/ComplicationTapBroadcastReceiver.kt [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+        // Retrieve data via SharedPreferences.
+        val preferenceKey = getPreferenceKey(provider, complicationId)
+        val sharedPreferences =
+            context.getSharedPreferences(COMPLICATION_PROVIDER_PREFERENCES_FILE_KEY, 0)
+        var value = sharedPreferences.getInt(preferenceKey, 0)
+
+        // Update data for complication.
+        value = (value + 1) % MAX_NUMBER
+        sharedPreferences.edit { putInt(preferenceKey, value) }
```
Comment (codingjeremy) [inline 3-trigger-updates/src/main/java/com/example/android/wearable/complicationsdataprovider/ComplicationTapBroadcastReceiver.kt]: `apply()`?

File: 4-long-data/src/main/java/com/example/android/wearable/complicationsdataprovider/ComplicationTapBroadcastReceiver.kt [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+        // Retrieve data via SharedPreferences.
+        val preferenceKey = getPreferenceKey(provider, complicationId)
+        val sharedPreferences =
+            context.getSharedPreferences(COMPLICATION_PROVIDER_PREFERENCES_FILE_KEY, 0)
+        var value = sharedPreferences.getInt(preferenceKey, 0)
+
+        // Update data for complication.
+        value = (value + 1) % MAX_NUMBER
+        sharedPreferences.edit { putInt(preferenceKey, value) }
```
Comment (codingjeremy) [inline 4-long-data/src/main/java/com/example/android/wearable/complicationsdataprovider/ComplicationTapBroadcastReceiver.kt]: `apply()`?

File: 5-ranged-data/src/main/java/com/example/android/wearable/complicationsdataprovider/ComplicationTapBroadcastReceiver.kt [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+        // Retrieve data via SharedPreferences.
+        val preferenceKey = getPreferenceKey(provider, complicationId)
+        val sharedPreferences =
+            context.getSharedPreferences(COMPLICATION_PROVIDER_PREFERENCES_FILE_KEY, 0)
+        var value = sharedPreferences.getInt(preferenceKey, 0)
+
+        // Update data for complication.
+        value = (value + 1) % MAX_NUMBER
+        sharedPreferences.edit { putInt(preferenceKey, value) }
```
Comment (codingjeremy) [inline 5-ranged-data/src/main/java/com/example/android/wearable/complicationsdataprovider/ComplicationTapBroadcastReceiver.kt]: `apply()`?

File: 6-all-data/src/main/java/com/example/android/wearable/complicationsdataprovider/ComplicationTapBroadcastReceiver.kt [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+        // Retrieve data via SharedPreferences.
+        val preferenceKey = getPreferenceKey(provider, complicationId)
+        val sharedPreferences =
+            context.getSharedPreferences(COMPLICATION_PROVIDER_PREFERENCES_FILE_KEY, 0)
+        var value = sharedPreferences.getInt(preferenceKey, 0)
+
+        // Update data for complication.
+        value = (value + 1) % MAX_NUMBER
+        sharedPreferences.edit { putInt(preferenceKey, value) }
```
Comment (codingjeremy) [inline 6-all-data/src/main/java/com/example/android/wearable/complicationsdataprovider/ComplicationTapBroadcastReceiver.kt]: `apply()`?

**Review by codingjeremy — COMMENTED** [review]
Overall, looks good, I just had one question on the `apply()` missing when pushing to SharedPreferences.

Also, feel free to remove extra steps so there is only start and complete in follow up PRs. :)

File: 1-base/src/main/java/com/example/android/wearable/complicationsdataprovider/ComplicationTapBroadcastReceiver.kt (hunk 1 above)
Comment (JolandaVerhoef) [reply 1-base/src/main/java/com/example/android/wearable/complicationsdataprovider/ComplicationTapBroadcastReceiver.kt]: Not when you use android-ktx. See https://developer.android.com/kotlin/ktx

**Comment by JolandaVerhoef** [comment]
@codingjeremy Can you re-review? The `edit {...}` is from `android-ktx` and doesn't need an `apply()` call.
I also took the liberty of removing the intermediary steps in this PR.

**Review by codingjeremy — APPROVED** [review]
Oh cool, didn't know that. :) 

LGTM, thanks!

**Pull request merged**